### PR TITLE
Feat/allow custom origins

### PR DESCRIPTION
**Pull request by msadura**
### <samp>🤖 Generated by Copilot at 568670c</samp>

This pull request introduces a new module `allowedOriginsCache` that caches the custom domains of the spaces and allows the websocket server to use them as allowed origins. This improves the security and performance of the websocket communication.

### WHY
We need a way to allow access to websocket server for custom domains. It is a matter of loading list of custom domains and using it for cors settings

### CODE REVIEW

File: lib/middleware/allowedOriginsCache.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+const DEFAULT_ORIGINS = ['.charmverse.co', '.charmverse.io'];
+
+class AllowedOriginsCache {
+  allowedOrigins: string[] | null = null;
+
+  constructor() {
+    this.refreshOrigins();
+  }
+
+  async refreshOrigins() {
```
Comment (msadura) [inline lib/middleware/allowedOriginsCache.ts:15]: in the future, this method can be called whenever we need to update custom origins list (i.e. when user saves custom domain)

File: background/websockets.ts [hunk 2]
```diff
@@ -22,10 +23,14 @@ const io = new Server(server, {
...
     credentials: true,
     origin: (requestOrigin, callback) => {
-      // support any subdomain for staging
-      if (requestOrigin?.endsWith('.charmverse.co') || requestOrigin?.endsWith('.charmverse.io')) {
+      if (isDevEnv) {
         callback(null, requestOrigin);
-      } else if (isDevEnv) {
+      } else if (
+        // custom origins did not load yet, allow all origins temporarily
+        !allowedOriginsCache.allowedOrigins ||
```
Comment (mattcasey) [inline background/websockets.ts:30]: can we actually just check for .charmverse.co/.io first? no need to wait for a cache for 99% of spaces

File: lib/middleware/allowedOriginsCache.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    const spaceCustomDomains = await prisma.space.findMany({
```
Comment (mattcasey) [inline lib/middleware/allowedOriginsCache.ts:16]: i think this should cache results one at a time - a map of promises, i guess. Loading all spaces will be a memory issue at some point

Comment (msadura) [reply lib/middleware/allowedOriginsCache.ts:16]: but I query only space with custom domains, we care only about those:
`customDomain: { not: null }`

**Comment by msadura** [comment]
> please change the logic to something like:
> 
> 1. check dev
> 2. check whitelist of charmverse domains
> 3. check if specific domain exists, and cache the result by domain
> 
> In the future we might add something like a TTL for each cache key, but since its just being whitelisted its probably not an issue

@mattcasey I ended up having a 5min cache - origin method is called on every page change (editor connect), so it feels better to have some sort of caching. LMK if you think we should still ignore caching.

File: background/websockets.ts [hunk 4]
```diff
@@ -21,14 +22,21 @@ const io = new Server(server, {
...
-    origin: (requestOrigin, callback) => {
+    origin: async (requestOrigin, callback) => {
       // support any subdomain for staging
       if (requestOrigin?.endsWith('.charmverse.co') || requestOrigin?.endsWith('.charmverse.io')) {
         callback(null, requestOrigin);
       } else if (isDevEnv) {
         callback(null, requestOrigin);
       } else {
-        callback(new Error('Not allowed by CORS'));
+        const isCustomOriginAllowed = await verifyCustomOrigin(requestOrigin);
```
Comment (mattcasey) [inline background/websockets.ts:32]: need to add a catch since i dont think origin is designed to handle a promise

**Pull request merged**